rust/rope/src/multiset.rs

```diff
@@ -0,0 +1,672 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A data structure for representing multi-subsets of sequences (typically strings).
+
+use std::cmp;
+
+// These two imports are for the `apply` method only.
+use tree::{Node, NodeInfo, TreeBuilder};
+use interval::Interval;
+use std::slice;
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+struct Segment {
+    len: usize,
+    count: usize,
+}
+
+/// Represents a multi-subset of a string, that is a subset where elements can
+/// be included multiple times. This is represented as each element of the
+/// string having a "count" which is the number of times that element is
+/// included in the set.
+///
+/// Internally, this is stored as a list of "segments" with a length and a count.
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct Subset {
+    /// Invariant, maintained by `SubsetBuilder`: all `Segment`s have non-zero
+    /// length, and no `Segment` has the same count as the one before it.
+    segments: Vec<Segment>,
+}
+
+#[derive(Default)]
+pub struct SubsetBuilder {
+    segments: Vec<Segment>,
+    total_len: usize,
+}
+
+impl SubsetBuilder {
+    pub fn new() -> SubsetBuilder {
+        SubsetBuilder::default()
+    }
+
+    /// Intended for use with `add_range` to ensure the total length of the
+    /// `Subset` corresponds to the document length.
+    pub fn pad_to_len(&mut self, total_len: usize) {
+        if total_len > self.total_len {
+            // might be able to extend last segment
+            if let Some(last) = self.segments.last_mut() {
+                if last.count == 0 {
+                    last.len += total_len - self.total_len;
+                    self.total_len = total_len;
+                    return;
+                }
+            }
+
+            self.segments.push(Segment {len: total_len - self.total_len, count: 0});
+            self.total_len = total_len;
+        }
+    }
+
+    /// Sets the count for a given range. This method must be called with a
+    /// non-empty range with beg not before the largest range or segment added
+    /// so far. Gaps will be filled with a 0-count segment.
+    pub fn add_range(&mut self, beg: usize, end: usize, count: usize) {
+        assert!(beg >= self.total_len, "ranges must be added in non-decreasing order");
+        // assert!(beg < end, "ranges added must be non-empty: [{},{})", beg, end);
+        if beg >= end {
+            return;
+        }
+        let len = end-beg;
+
+        // merge into previous segment if possible
```

Comment:
I think you could just get rid of this, the following two calls to `push` take care of this case.